### PR TITLE
AlphaZero

### PR DESCRIPTION
**Pull request by AntoineDubois**
This pull request introduces the implementation of the *AlphaZero algorithm in `mlpack`, along with several supporting components and modifications. The following major additions and improvements are included:

# Key Contributions:
1. *RandomCategorical*: 
- Added a new random generator for sampling from categorical distributions. This is useful for probabilistic policy selection, particularly in reinforcement learning tasks like AlphaZero.
2. *CrossEntropy*: 
- Implemented a new cross-entropy loss function designed to compute the loss between two distributions.
- Removed the misleading alias of the binary cross-entropy loss (`BCELoss`), which could cause confusion in certain contexts. This change improves clarity and precision in the library's functionality.
3. *AlphaZero*:
- Complete implementation of the AlphaZero algorithm, which includes both training and inference capabilities.
- A comprehensive tutorial is provided to demonstrate how to use AlphaZero in practice. This includes detailed steps for training, policy updates, and using the model for decision-making.
4. *RandomReplay (Significant Modification)*:
- Major enhancements to the existing RandomReplay class to allow the storage and retrieval of policy vectors, which is essential for AlphaZero's experience replay mechanism. This modification enables more flexible replay capabilities.

# Additional Information:
1. *Documentation and Tutorials*: 
- The PR includes a detailed tutorial on how to use AlphaZero within mlpack. This tutorial covers setup and training, making it easier for users to get started with the algorithm.
2. *Backward Compatibility*:
- Where applicable, care has been taken to ensure backward compatibility, particularly with the modifications to RandomReplay and the removal of the misleading alias in `BCELoss`.
---
This pull request aims to integrate a robust and flexible implementation of AlphaZero, accompanied by essential enhancements in loss functions and replay buffers. These changes are critical for enabling reinforcement learning workflows within `mlpack` and expanding the library's capabilities in this domain.
---
# How to Review:
- The changes can be reviewed incrementally starting with the `RandomCategorical` class and `CrossEntropy` function.
- Special attention should be given to the modifications in `RandomReplay`, as it is a core component for the AlphaZero algorithm's experience replay mechanism.

### CODE REVIEW

**Comment by shrit** [comment]
Bonjour Antoine,

Thank you for adding this, I had a quick look at the code, could you verify the following:

1. Please be sure that your PR compiles locally on your device with the tests, and be sure that the tests pass.
2. Be sure to use mlpack style use clang_format from one of the open PR to fix it,
3. Please templatize the use of armadillo, check other methods to get some inspiration
4. Ping me and @rcurtin  when you finish all of the above.

Merci

File: CMake/ConfigureCrossCompile.cmake [hunk 1]
```diff
@@ -28,7 +28,6 @@ macro(search_openblas version)
     get_deps(https://github.com/xianyi/OpenBLAS/releases/download/v${version}/OpenBLAS-${version}.tar.gz OpenBLAS OpenBLAS-${version}.tar.gz)
     if (NOT MSVC)
       if (NOT EXISTS "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
-        set(ENV{COMMON_OPT} "${CMAKE_OPENBLAS_FLAGS}") # Pass our flags to OpenBLAS
```
Comment (shrit) [inline CMake/ConfigureCrossCompile.cmake:31]: Why this is being removed ?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+# Default target executed when no arguments are given to make.
+default_target: all
+.PHONY : default_target
+
+# Allow only one "make -f Makefile2" at a time, but pass parallelism.
+.NOTPARALLEL:
+
+#=============================================================================
+# Special targets provided by cmake.
```
Comment (shrit) [inline Makefile:12]: Why this file being added ?

**Review by shrit — CHANGES_REQUESTED** [review]
I really do not like that part of the code (hopefully not all of it) is being generated by ChatGPT, if you are here to spam people, then please go to another project and start your spamming career there. We have been blocking people that are being identified doing this kind of silly behavior, while I appreciate you opened an issue last year and you are interested in using the library, you should know that we are interested in having genuine implementation. Consider this as a fair warning, before reporting your account to github

File: src/mlpack/core/math/random.hpp [hunk 3]
```diff
@@ -21,8 +21,7 @@ namespace mlpack {
 // seeds do not cause each thread's RNG to have the exact same output.
 // Therefore, we assign a different offset for each thread, and add this offset
 // whenever we set a seed with RandGen() or RandomSeed().
-inline size_t RandGenSeedOffset()
-{
+inline size_t RandGenSeedOffset() {
```
Comment (rcurtin) [inline src/mlpack/core/math/random.hpp:24]: Not sure why you changed the style in these files?

File: src/mlpack/core/math/random.hpp [hunk 4]
```diff
@@ -162,11 +145,26 @@ inline double RandNormal()
...
+inline double RandNormal(const double mean, const double stddev) {
   return stddev * RandNormalDist()(RandGen()) + mean;
 }
 
+/**
+ * Generates random samples from a categorical distribution based
+ *
+ * @param weights arma::Col<double> of weights summing to 1.
+ */
+inline size_t RandCategorical(arma::vec &weights) {
```
Comment (rcurtin) [inline src/mlpack/core/math/random.hpp:157]: Realistically you could use `DiscreteDistribution` here instead of adding something new.

File: Makefile (hunk 2 above)
Comment (rcurtin) [reply Makefile:12]: I think this was an accidental checkin, but it suggests that you are using CMake without a separate build directory (i.e. an in-source build) which is generally not a great idea; better to create a `build/` directory and keep all build artifacts in there.

**Review by rcurtin — COMMENTED** [review]
I haven't reviewed the code here because I found the LLM-generated summary very difficult to understand.  For instance this sentence says nothing:

> Removed the misleading alias of the binary cross-entropy loss (BCELoss), which could cause confusion in certain contexts. This change improves clarity and precision in the library's functionality.

What is the confusion and what are the certain contexts?  I can't evaluate whether or not the change is even a good one if I can't understand what the criticism of the current implementation is.

I have a couple high-level comments on code reuse, but to be honest, this is a very large feature and I am not the right person to review it (my expertise is not RL).  I might suggest smaller changes to the codebase first, that are more in line with the [vision document](https://www.mlpack.org/vision.pdf).  If I was going to review it in depth, I would probably ask for more tests.

Do be careful using LLMs: from our perspective, we don't know if you used an LLM for the whole thing, or just the description, or what.  That makes it hard for me as a reviewer to trust the code; mlpack contains lots of tricky numerical linear algebra code, and it's really easy to go wrong here in subtle and unexpected ways.  When a contributor uses an LLM tool to write the code, as a reviewer, it's tough for me to trust the code at a baseline level (and even beyond that, I wouldn't want to spend hours reviewing something that an LLM wrote in seconds that wasn't thoroughly double-checked, it's not a good use of maintainer time).

In any case, I hope the comments are helpful here, in some way or another.

**Comment by AntoineDubois** [comment]
Omar,
I understand that you are concerned, but I want to clarify that implementing AlphaZero in mlpack has required significant work from my part, and I have been committed to contributing to mlpack. I was surprised and concerned by your message threatening to report my account to GitHub.
My intention has never been to spam or corrupt mlpack. While I used AI to proofread the pull request report, the code itself is the result of thorough work, and, I believe, is sound.
I am happy to address the feedback to the pull request to ensure it meets the project’s standards.
I hope we can move forward constructively,
Antoine Dubois

**Comment by shrit** [comment]
@AntoineDubois Okay understood, sorry for my comment, you should be able to handle the comments that we have left, please let us know if you need any help
